### PR TITLE
Implement MRS cross-artifact algorithm in MRS straylight correction

### PR DESCRIPTION
**Pull request by drlaw1558**
Resolves [JP-2630](https://jira.stsci.edu/browse/JP-2630)

**Description**

This PR implements the cross-artifact model in place of the previous polynomial model for the MRS straylight correction algorithm.  For speed purposes it is implemented as python-wrapped C, although a pure-python routine is still available for test purposes.

There is still some code cleanup necessary, but opening this PR now to start getting feedback.

This PR also implements some changes to the mrsxartcorr data model with respect to the previous prototype file delivered to CRDS.

Checklist
- [x] Tests
- [x] Documentation
- [x] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

**Comment by drlaw1558** [comment]
Clearly a lot of tests need to be updated...

File: jwst/straylight/straylight.py [hunk 1]
```diff
@@ -1,290 +1,180 @@
...
-    # normalize straylight flux by weights
-    norm_conv = convolve(mask, w)
-
-    astropy_conv /= norm_conv
-
+    # Which channel/band are we using?
+    channel = input_model.meta.instrument.channel
+    band = input_model.meta.instrument.band
+
+    # Deal with normal cases only, we won't apply to cross-dichroic cases for now
```
Comment (hbushouse) [inline jwst/straylight/straylight.py:129]: Should there be a final else clause to act as a trap to handle cross-dichroic cases? i.e. issue a warning and skip the step

File: jwst/straylight/straylight_step.py [hunk 2]
```diff
@@ -14,91 +14,36 @@ class StraylightStep (Step):
...
 
-    spec = """
-         method = option('Nearest','ModShepard',default='ModShepard') #Algorithm method
-         roi = integer(2, 1024, default = 50) # Region of interest given as even integer
-         power = float(0.1, 5, default = 1.0) # Power of weighting function
-
-    """
-
-    reference_file_types = ['regions']
+    reference_file_types = ['mrsxartcorr']
```
Comment (hbushouse) [inline jwst/straylight/straylight_step.py:17]: Should update the doc page we have somewhere in jwst/docs/jwst that lists the ref file types used by each step

**Review by hbushouse — COMMENTED** [review]
Looks great. Just a few minor comments.

**Comment by hbushouse** [comment]
Failures in the CI test that builds the docs is due to lack of doc file comparable to https://raw.githubusercontent.com/spacetelescope/jwst/master/docs/jwst/references_general/regions_reffile.rst. Need one for the new mrsxart ref file.

File: CHANGES.rst [hunk 3]
```diff
@@ -37,6 +37,7 @@ straylight
 ----------
 
 - Add a check that input data is IFUImageModel [#6861]
+- Update straylight algorithm to use cross-artifact model [#6873]
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply CHANGES.rst:41]: Done

File: jwst/straylight/straylight.py [hunk 4]
```diff
@@ -1,290 +1,180 @@
...
-    function. The inverse distance weighting is determined by module
-    ``shepard_2d_kernel(roi,power)``.
+def correct_xartifact(input_model, modelpars):
+    """
+    Corrects the MIRI MRS data for 'straylight' produced by the cross-artifact.
 
     Parameters
     ----------
-    input_model : `~jwst.datamodels.IFUImageModel`
+    input_model: `~jwst.datamodels.IFUImageModel`
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply jwst/straylight/straylight.py]: Done

File: jwst/straylight/straylight.py (hunk 1 above)
Comment (drlaw1558) [reply jwst/straylight/straylight.py:129]: I think a final else clause might catch things I didn't intend given how the if statements are set up, but I've added a default setting before the if statements, an informative message if that doesn't get changed by the if statements, and the try/except should catch any problems.

File: jwst/straylight/straylight_step.py (hunk 2 above)
Comment (drlaw1558) [reply jwst/straylight/straylight_step.py:17]: Found some pages at docs/jwst/references_general/ that looked like it could use some doc pages by analogy to other steps, so created them.  Let me know if anywhere else specific is missing.

File: jwst/straylight/src/calc_xart.c [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  // This is how big the output vector needs to be
+  int npt = xsize_det * ysize_det;
+
+  // allocate memory to hold output 
+  if (alloc_xart_arrays(npt, &fluxv)) return 1;
+
+  for (j = 0; j < ysize_det; j++) {
+    for (i = imin; i < imax; i++) {
+      for (k = 0; k < xsize_det; k++) {
+        fluxv[xsize_det * j + k] += (fimg[j * xsize_det + i] * lor_amp[j] * gamma[j] * gamma[j])
```
Comment (jemorrison) [inline jwst/straylight/src/calc_xart.c:110]: Question: the array fluxv is allocated memory for 'npt' elements =xsize_det * ysize_det . But in the for loop it looks like fluxv will contain 'xsize_det * ysize_det + imax' elements.  I am sure you have tested this and if I was correct it would give you an error - so I must be missing something.

Comment (drlaw1558) [reply jwst/straylight/src/calc_xart.c:110]: I think this is ok; fluxv has size xsize_det * ysize_det.  It gets populated into element xsize_det*j + k

The maximum value of j is ysize_det-1 and the maximum value of k is xsize_det -1

So the maximum element should be xsize_det * (ysize_det-1) + xsize_det - 1 = xsize_det*ysize_det - xsize_det + xsize_det - 1 = xsize_det * ysize_det - 1

Likewise since imax is never bigger than 1031=xsize_det-1

j*xsize_det + i can never be more than (ysize_det-1)*xsize_det + xsize_det-1 = xsize*det*ysize_det - xsize_det + xsize_det -1 = xsize_det*ysize_det - 1

Hasn't crashed in running it pretty extensively.

**Review by jemorrison — APPROVED** [review]
That was a total re-write of the step- which makes sense cause we now know the cause of the straylight. It all seems to make sense  but I don't think  we will find bugs until this is  run on all sorts of data - there we will probably have edge cases and such we need to fix. We need to add to the unit tests.

File: docs/jwst/straylight/reference_files.rst [hunk 6]
```diff
@@ -1,11 +1,7 @@
...
-different throughput levels ranging from 10% - 90%.  While ``assign_wcs``
-uses a fairly exclusive slice mask (selecting only pixels with high
-throughput for science analysis), the ``straylight`` step uses a very
-inclusive 20% threshold to define slice pixels and obtain a cleaner
-sample of inter-slice pixels.
+The ``straylight`` step uses the :ref:`MRSXARTCORR <mrsxartcorr_reffile>` reference
+file, which stores vectors describing the appropriate cross-artifact convolution kernel
+for each MRS band.  In Channel 1 these vectors include power in a broad Lorentzian core
+plus a pair of double-Gaussian profiles.  In Channels 2 and 3 these vectors include only
+power in the broad Lorentzian, while in Channel 4 there is no correction.
```
Comment (hbushouse) [inline docs/jwst/straylight/reference_files.rst:7]: The "mrsxartcorr_reffile.inc" file (defined above) is not getting included anywhere from any of the other .rst files, hence it's not getting compiled at all and not included in the doc build. Typically that .inc file is loaded somewhere in the "reference_files.rst" doc that describes the reference file. So just add a line here at the end of this file:
`.. include:: ../references_general/mrsxartcorr_reffile.inc`

If you want, you could even move this little paragraph of text to the top of the mrsxartcorr_reffile.inc and reduce this file to only contain:
```
Reference Files
===========
The ``straylight`` step uses a MRSXARTCORR reference file.

.. include:: ../references_general/mrsxartcorr_reffile.inc
```

File: docs/jwst/references_general/mrsxartcorr_reffile.inc [hunk 7]
```diff
@@ -0,0 +1,59 @@
+.. _mrsxartcorr_reffile:
+
+MRSXARTCORR reference file
+----------------------
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply docs/jwst/references_general/mrsxartcorr_reffile.inc]: done

File: docs/jwst/references_general/mrsxartcorr_reffile.inc [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+The MRSXARTCORR reference file contains parameter values used to model and subtract
+the cross-artifact in the straylight step.
+
+.. include:: ../references_general/mrsxartcorr_selection.inc
+
+.. include:: ../includes/standard_keywords.inc
+
+Type Specific Keywords for MRSXARTCORR
++++++++++++++++++++++++++++++++++++++
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply docs/jwst/references_general/mrsxartcorr_reffile.inc]: done

File: docs/jwst/references_general/mrsxartcorr_selection.inc [hunk 9]
```diff
@@ -0,0 +1,13 @@
+.. _mrsxartcorr_selectors:
+
+Reference Selection Keywords for MRSXARTCORR
+++++++++++++++++++++++++++++++++++++++++
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply docs/jwst/references_general/mrsxartcorr_selection.inc]: done

**Comment by hbushouse** [comment]
So close!  All the docs for the step and the ref file are now building correctly. The only thing still missing is to update the references to the kind of ref file used in the straylight step in the file https://github.com/spacetelescope/jwst/blob/master/docs/jwst/references_general/references_general.rst

I thought I had seen that in an earlier commit, but maybe not.

**Comment by drlaw1558** [comment]
Nope, didn't know about the references_general.rst; modified now!

**Review by hbushouse — APPROVED** [review]
Looks good!
